File: packages/sdk-rust/src/threads.rs
use crate::client::TaskForceAI;
use crate::error::TaskForceAIError;
use crate::validation::{
    validate_thread, validate_thread_list, validate_thread_messages, validate_thread_run,
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Represents a conversation thread.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Thread {
    pub id: i64,
    pub timestamp: String,
    pub user_input: String,
    pub result: String,
    pub execution_time: i64,
    pub model: String,
    pub agent_count: i64,
    pub sources: Option<Vec<serde_json::Value>>,
    #[serde(rename = "agentStatuses")]
    pub agent_statuses: Option<Vec<serde_json::Value>>,
    #[serde(rename = "toolEvents")]
    pub tool_events: Option<Vec<serde_json::Value>>,
}

/// Represents a message within a thread.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessage {
    pub id: i64,
    pub thread_id: i64,
    pub role: String, // "user" or "assistant"
    pub content: String,
    #[serde(default)]
    pub message_id: Option<String>,
    #[serde(default)]
    pub is_agent_status: bool,
    #[serde(default)]
    pub elapsed_seconds: Option<f64>,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub sources: Option<serde_json::Value>,
    #[serde(default)]
    pub tool_events: Option<serde_json::Value>,
    #[serde(default)]
    pub agent_statuses: Option<serde_json::Value>,
    #[serde(default)]
    pub updated_at: Option<String>,
    #[serde(default)]
    pub rating: i32,
}

/// Options for creating a thread.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CreateThreadOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing)]
    pub messages: Option<Vec<ThreadMessage>>,
    #[serde(skip_serializing)]
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

/// Response containing a list of threads.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadListResponse {
    pub conversations: Vec<Thread>,
    pub total: i64,
    pub limit: i64,
    pub offset: i64,
    pub has_more: bool,
}

/// Response containing messages from a thread.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessagesResponse {
    pub messages: Vec<ThreadMessage>,
    #[serde(default)]
    pub truncated: bool,
}

/// Options for running a prompt in a thread.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThreadRunOptions {
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,
    #[serde(skip_serializing)]
    pub options: Option<HashMap<String, serde_json::Value>>,
}

/// Response from running in a thread.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadRunResponse {
    #[serde(rename = "taskId")]
    pub task_id: String,
    pub status: String,
}

impl TaskForceAI {
    /// Creates a new conversation thread.
    pub async fn create_thread(
        &self,
        options: Option<CreateThreadOptions>,
    ) -> Result<Thread, TaskForceAIError> {
        let body = options
            .map(serde_json::to_value)
            .transpose()?
            .unwrap_or_else(|| serde_json::json!({}));

        let thread: Thread = self
            .request(reqwest::Method::POST, "/threads", Some(body))
            .await?;
        validate_thread(&thread, "thread")?;
        Ok(thread)
    }

    /// Retrieves a list of threads.
    pub async fn list_threads(
        &self,
        limit: i32,
        offset: i32,
    ) -> Result<ThreadListResponse, TaskForceAIError> {
        let path = format!("/threads?limit={}&offset={}", limit, offset);
        let response: ThreadListResponse = self.request(reqwest::Method::GET, &path, None).await?;
        validate_thread_list(&response)?;
        Ok(response)
    }

    /// Retrieves a specific thread by ID.
    pub async fn get_thread(&self, thread_id: i64) -> Result<Thread, TaskForceAIError> {
        let path = format!("/threads/{}", thread_id);
        let thread: Thread = self.request(reqwest::Method::GET, &path, None).await?;
        validate_thread(&thread, "thread")?;
        Ok(thread)
    }

    /// Returns an error because deleting threads is not currently supported by the Developer API.
    pub async fn delete_thread(&self, thread_id: i64) -> Result<(), TaskForceAIError> {
        Err(TaskForceAIError::Other(format!(
            "delete_thread is not supported by the current Developer API (thread_id={})",
            thread_id
        )))
    }

    /// Retrieves messages from a thread.
    pub async fn get_thread_messages(
        &self,
        thread_id: i64,
        limit: i32,
        offset: i32,
    ) -> Result<ThreadMessagesResponse, TaskForceAIError> {
        let path = format!(
            "/threads/{}/messages?limit={}&offset={}",
            thread_id, limit, offset
        );
        let response: ThreadMessagesResponse =
            self.request(reqwest::Method::GET, &path, None).await?;
        validate_thread_messages(&response)?;
        Ok(response)
    }

    /// Submits a prompt within a thread context.
    pub async fn run_in_thread(
        &self,
        thread_id: i64,
        options: ThreadRunOptions,
    ) -> Result<ThreadRunResponse, TaskForceAIError> {
        if options.prompt.trim().is_empty() {
            return Err(TaskForceAIError::EmptyPrompt);
        }

        let path = format!("/threads/{}/runs", thread_id);
        let body = serde_json::to_value(options)?;

        let response: ThreadRunResponse = self
            .request(reqwest::Method::POST, &path, Some(body))
            .await?;
        validate_thread_run(&response)?;
        Ok(response)
    }
}

File: packages/sdk-go/threads.go
package taskforceai

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// Thread represents a conversation thread.
type Thread struct {
	ID            int              `json:"id"`
	Timestamp     string           `json:"timestamp"`
	UserInput     string           `json:"user_input"`
	Result        string           `json:"result"`
	ExecutionTime int              `json:"execution_time"`
	Model         string           `json:"model"`
	AgentCount    int              `json:"agent_count"`
	Sources       []map[string]any `json:"sources"`
	AgentStatuses []map[string]any `json:"agentStatuses"`
	ToolEvents    []map[string]any `json:"toolEvents"`
}

// ThreadMessage represents a message within a thread.
type ThreadMessage struct {
	ID             int             `json:"id"`
	ThreadID       int             `json:"thread_id"`
	Role           string          `json:"role"` // "user" or "assistant"
	Content        string          `json:"content"`
	MessageID      string          `json:"message_id,omitempty"`
	CreatedAt      string          `json:"created_at,omitempty"`
	UpdatedAt      string          `json:"updated_at,omitempty"`
	Error          *string         `json:"error,omitempty"`
	Sources        json.RawMessage `json:"sources,omitempty"`
	ToolEvents     json.RawMessage `json:"tool_events,omitempty"`
	AgentStatuses  json.RawMessage `json:"agent_statuses,omitempty"`
	IsAgentStatus  bool            `json:"is_agent_status"`
	ElapsedSeconds *float64        `json:"elapsed_seconds,omitempty"`
	Rating         int32           `json:"rating"`
}

// CreateThreadOptions contains options for creating a thread.
type CreateThreadOptions struct {
	Title    string          `json:"title,omitempty"`
	Messages []ThreadMessage `json:"messages,omitempty"`
	Metadata map[string]any  `json:"metadata,omitempty"`
}

// ThreadListResponse contains a list of threads.
type ThreadListResponse struct {
	Conversations []Thread `json:"conversations"`
	Total         int      `json:"total"`
	Limit         int      `json:"limit"`
	Offset        int      `json:"offset"`
	HasMore       bool     `json:"has_more"`
}

// ThreadMessagesResponse contains messages from a thread.
type ThreadMessagesResponse struct {
	Messages  []ThreadMessage `json:"messages"`
	Truncated bool            `json:"truncated,omitempty"`
}

// ThreadRunOptions contains options for running a prompt in a thread.
type ThreadRunOptions struct {
	Prompt  string         `json:"prompt"`
	ModelID string         `json:"modelId,omitempty"`
	Stream  bool           `json:"stream,omitempty"`
	Options map[string]any `json:"-"` // Deprecated: unsupported by the current Developer API.
}

// ThreadRunResponse contains the result of running in a thread.
type ThreadRunResponse struct {
	TaskID string `json:"taskId"`
	Status string `json:"status"`
}

// CreateThread creates a new conversation thread.
func (c *Client) CreateThread(ctx context.Context, opts *CreateThreadOptions) (*Thread, error) {
	body := map[string]any{}
	if opts != nil {
		if opts.Title != "" {
			body["title"] = opts.Title
		}
	}

	return requestDecoded(c, ctx, http.MethodPost, "/threads", body, "create thread", 0, "thread", validateThread)
}

// ListThreads retrieves a list of threads.
func (c *Client) ListThreads(ctx context.Context, limit, offset int) (*ThreadListResponse, error) {
	path := fmt.Sprintf("/threads?limit=%d&offset=%d", limit, offset)

	return requestDecoded(c, ctx, http.MethodGet, path, nil, "list threads", http.StatusOK, "thread list", validateThreadList)
}

// GetThread retrieves a specific thread by ID.
func (c *Client) GetThread(ctx context.Context, threadID int) (*Thread, error) {
	path := fmt.Sprintf("/threads/%d", threadID)

	return requestDecoded(c, ctx, http.MethodGet, path, nil, "get thread", http.StatusOK, "thread", validateThread)
}

// DeleteThread is currently unsupported because the Developer API has no thread delete endpoint.
func (c *Client) DeleteThread(ctx context.Context, threadID int) error {
	_ = ctx
	return fmt.Errorf(
		"DeleteThread is not supported by the current Developer API (threadID=%d)",
		threadID,
	)
}

// GetThreadMessages retrieves messages from a thread.
func (c *Client) GetThreadMessages(ctx context.Context, threadID int, limit, offset int) (*ThreadMessagesResponse, error) {
	path := fmt.Sprintf("/threads/%d/messages?limit=%d&offset=%d", threadID, limit, offset)

	return requestDecoded(c, ctx, http.MethodGet, path, nil, "get thread messages", http.StatusOK, "thread messages", validateThreadMessages)
}

// RunInThread submits a prompt within a thread context.
func (c *Client) RunInThread(ctx context.Context, threadID int, opts ThreadRunOptions) (*ThreadRunResponse, error) {
	if opts.Prompt == "" {
		return nil, fmt.Errorf("prompt is required")
	}

	path := fmt.Sprintf("/threads/%d/runs", threadID)
	body := map[string]any{
		"prompt": opts.Prompt,
	}
	if opts.ModelID != "" {
		body["modelId"] = opts.ModelID
	}
	if opts.Stream {
		body["stream"] = true
	}

	return requestDecoded(c, ctx, http.MethodPost, path, body, "run in thread", 0, "thread run", validateThreadRun)
}
